crates/oxc_span/src/source_type.rs

```diff
@@ -166,25 +174,222 @@ impl SourceType {
             .ok_or_else(|| {
                 let path = path.as_ref().to_string_lossy();
                 UnknownExtension(
-                    format!("Please provide a valid file extension for {path}: .js, .mjs, .jsx or .cjs for JavaScript, or .ts, .mts, .cts or .tsx for TypeScript"),
+                    format!("Please provide a valid file extension for {path}: .js, .mjs, .jsx, .vue or .cjs for JavaScript, or .ts, .mts, .cts, .vue or .tsx for TypeScript"),
                 )
             })?;
 
+        let (is_ts_in_special_file, range) = Self::may_parser_special_extension(&path);
+
         let is_definition_file = file_name.ends_with(".d.ts")
             || file_name.ends_with(".d.mts")
             || file_name.ends_with(".d.cts");
 
         let language = match extension {
             "js" | "mjs" | "cjs" | "jsx" => Language::JavaScript,
             "ts" | "mts" | "cts" | "tsx" => Language::TypeScript { is_definition_file },
+            "vue" => {
+                if is_ts_in_special_file.unwrap_or(false) {
+                    Language::TypeScript { is_definition_file }
+                } else {
+                    Language::JavaScript
+                }
+            }
             _ => unreachable!(),
         };
 
         let variant = match extension {
             "js" | "mjs" | "cjs" | "jsx" | "tsx" => LanguageVariant::Jsx,
+            "vue" => LanguageVariant::Vue,
             _ => LanguageVariant::Standard,
         };
 
-        Ok(Self { language, module_kind: ModuleKind::Module, variant, always_strict: false })
+        Ok(Self { language, module_kind: ModuleKind::Module, variant, always_strict: false, range })
+    }
+    pub fn range(&self) -> Option<Span> {
+        self.range
+    }
+    fn may_parser_special_extension<P: AsRef<Path>>(path: P) -> (Option<bool>, Option<Span>) {
+        let default_ret = (None, None);
+        let Ok(content) = std::fs::read_to_string(path.as_ref()) else { return default_ret };
```

Comment:
I don't really like here, it's means we need to read the content of Vue file twice. One here, another is in `process_path`
https://github.com/oxc-project/oxc/blob/23227305327b4a731095e3ac9641c3a63ee121d4/crates/oxc_linter/src/service.rs#L142-L153

How should I can improve it 😿